Login workflow

larkox opened the pull request:
#### Summary
Added a first approach to the login workflow defined [here](https://www.figma.com/file/OIGiBfSeZA7EQO2HRXpmPt/MS-Calendar?node-id=0%3A1). The current solution is far from optimal, but allows us to get a better understanding of the different moving parts of the solution.

I will keep working to improve the different parts, both in how to present the data, and the underlying architecture. Here is the current look of a full login process.

![image](https://user-images.githubusercontent.com/1933730/75814048-5ff8bf80-5d91-11ea-9bdb-5fd6df3524d4.png)

larkox commented:
Here I leave some screenshots of the current behaviour:
Ephemeral:
![image](https://user-images.githubusercontent.com/1933730/76399778-82856c80-637f-11ea-9e69-ed9875a32e8d.png)
Bot messages step 1:
![image](https://user-images.githubusercontent.com/1933730/76399862-a8ab0c80-637f-11ea-8f13-d94c4e1b9935.png)
Bot messages step 2:
![image](https://user-images.githubusercontent.com/1933730/76399934-caa48f00-637f-11ea-9d29-4572ef8f8da3.png)
Bot messages step 3:
![image](https://user-images.githubusercontent.com/1933730/76400085-093a4980-6380-11ea-93db-4eb5605c21a4.png)
Bot messages step 4:
![image](https://user-images.githubusercontent.com/1933730/76399996-e1e37c80-637f-11ea-92b9-a60ad1583054.png)

levb commented on server/store/welcome_store.go:
```diff
@@ -0,0 +1,36 @@
+package store
```
Why did you choose not to embed this data in the User records?

larkox replied:
As spoken through zoom, this data (the id of the post where the connect link is) is created before the User even exist. Therefore, it cannot be embed in the User records.

larkox commented on server/mscalendar/calendar.go:
```diff
@@ -55,7 +55,7 @@ func (m *mscalendar) CreateEvent(user *User, event *remote.Event, mattermostUser
 		if err != nil {
 			if err.Error() == "not found" {
-				err = m.Poster.DM(mattermostUserID, "You have been invited to an MS office calendar event but have not linked your account.  Feel free to join us by connecting your www.office.com using `/mscalendar connect`")
+				_, err = m.Poster.DM(mattermostUserID, "You have been invited to an MS office calendar event but have not linked your account.  Feel free to join us by connecting your www.office.com using `/mscalendar connect`")
```
Not on the scope of this PR (this change is just to comply with the new interface of  `Poster.DM`), but I think I could add it without much problem.

larkox commented on server/mscalendar/mscalendar.go:
```diff
@@ -18,6 +18,7 @@ type MSCalendar interface {
 	EventResponder
 	Subscriptions
 	Users
+	Welcome(userID string) error
```
(1/5) MSCalendar is not a welcomer, the bot in the dependencies is. But MSCalendar must know how to greet a new user, to be used on the commands, so that is why we expose that function on the interface.

levb replied:
Dunno, I'd advocate composing the entire interface in. What's the harm? Polluting MSCalendar namespace?

larkox commented:
@levb After our conversation yesterday, I could not stop thinking about making the flow more detached from the welcome logic, so I made a refactoring in this line. Now we can add any steps we want as long as they are Yes/No questions.

With the current implementation, we declare our flow with its steps (an implementation of Flow interface) and our flowStore (an implementation of FlowStore). With that, Welcomer still handles the connection, but all the flow of Yes/No questions is handled by the Flow.

larkox commented on Makefile:
```diff
@@ -98,6 +98,7 @@ mock:
 ifneq ($(HAS_SERVER),)
 	go install github.com/golang/mock/mockgen
 	mockgen -destination server/mscalendar/mock_mscalendar/mock_mscalendar.go github.com/mattermost/mattermost-plugin-mscalendar/server/mscalendar MSCalendar
+	mockgen -destination server/mscalendar/mock_mscalendar/mock_welcomer.go github.com/mattermost/mattermost-plugin-mscalendar/server/mscalendar Welcomer
```
In this case, I mocked the Welcomer because the test was calling `AfterSuccessfullyConnect`, which makes several operations that made the test less unitary and more confusing. Therefore, instead of using the "proper welcomer" already mocked on the MSCalendar, I preferred to mock it, and at some point maybe test it in a more unitary way.

mickmister commented on server/mscalendar/welcomer.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+func (bot *MSCalendarBot) AfterSuccessfullyConnect(userID, userLogin string) error {
+	postID, _ := bot.Store.DeleteUserWelcomePost(userID)
```
Why ignore the error here?

larkox replied:
(1/5) I consider this error not relevant. Among other things, the error may come because there is no Welcome Post stored. My only concern is whether I get a valid postID. I could log the error just in case, but after getting to this point, I do not think we should make the function fail due to a database inconsistency.

mickmister commented on server/mscalendar/welcomer.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+func (bot *MSCalendarBot) AfterDisconnect(userID string) error {
+	bot.Cancel(userID)
```
Ignoring error here too

mickmister commented on server/utils/bot/bot.go:
```diff
@@ -14,10 +15,12 @@ type Bot interface {
 	Poster
 	Logger
 	Admin
+	Flower
```
When I read this it makes me think of the plant flower :rose:. Not sure what else to name this though

larkox replied:
Exactly the same feeling here. But if something that posts is a poster, and something that logs is a logger, something that flows...

mickmister commented on server/utils/flow/handler.go:
```diff
@@ -0,0 +1,81 @@
+	err = fh.store.SetProperty(mattermostUserID, property, value)
+	if err != nil {
+		fmt.Println("There has been a problem setting the property, err=" + err.Error())
+		http.Error(w, "There has been a problem setting the property, err="+err.Error(), http.StatusInternalServerError)
```
I suggest having `fh` take a `Logger` so we can log this properly. Also put the error string in a var in the line above to be used in `Logger.Errorf` and `http.Error`, to avoid duplication.

larkox replied:
This was actually some lines for myself for debugging that slipped into the commit. Would you still like these errors to be logged in the server?

mickmister replied:
It doesn't seem too critical to me. 0/5 remove the debug print

mickmister commented on server/utils/bot/flower.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+func (bot *bot) setFlowStep(userID string, step int) error {
+	// TODO: Use KVStore for HA
```
I assume this will be covered in this PR? I'd be fine with deferring to a ticket if you want to get this merged sooner.

larkox replied:
My plan is to have this covered in this PR, but I wanted to confirm first that the new shape of the flow was interesting. I will take a look to have this done tomorrow.

levb commented on server/command/connect_test.go:
```diff
@@ -38,9 +38,11 @@ func TestConnect(t *testing.T) {
 			},
-			expectedOutput: "[Click here to link your Microsoft Calendar account.](http://localhost/oauth2/connect)",
-			expectedError:  "",
+			expectedOutput: `Welcome to the Microsoft Calendar Bot.
```
We should review the consolidated messages with @aaronrothschild and @asaadmahmood - let's focus on making sure they're all "in one place"?

levb commented on server/mscalendar/oauth2_test.go:
```diff
@@ -1,4 +1,4 @@
-package mscalendar
+package mscalendar_test
```
Maybe a confused question, but if it's a separate package, wouldn't `./test` be a better place?

larkox replied:
This is a workaround that Go has for cycles in the tests. Conceptually, it is still a test of mscalendar, but due to the import cycles, it cannot be part of mscalendar package. With this naming, the compiler consider this a separate package, which solves the import cycles.

levb replied:
^^ seems fishy to me, we should be well-clear of cyclical dependencies. I'm ok with the PR as is, but would like to understand this better, offline?

larkox replied:
Solved by creating the mock of the welcomer in a different package.

levb commented on server/utils/flow/flow.go:
```diff
@@ -0,0 +1,116 @@
+	}
+}
+
+func (s *step) PostSA(flowHandler string, i int) *model.SlackAttachment {
```
I know it's verbose, but 2/5 `SlackAttachment` please... (grep-ability)

larkox commented on server/mscalendar/welcomer.go:
```diff
@@ -0,0 +1,266 @@
+}
+
+func (bot *MSCalendarBot) AfterSuccessfullyConnect(userID, userLogin string) error {
+	user, err := bot.Store.LoadUser(userID)
```
I understand the benefit and how to do it, but I think with the new shape of the program this is not necessary. Could you confirm?

larkox commented:
With current commit, all previous comments but using KVStore in bot flower should be addressed.

mickmister commented:
@larkox Do you think it makes sense to ask the user when they want their daily summary during the sign-up flow?

larkox commented:
@mickmister Since it is not a Yes/no question, more changes should be made, but should be doable. The problem is how to show it. The best (and practically only) way to do it with the tools we have, would be a select, but since we are allowing 15 minutes intervals, that would be a select with almost 100 options. Not sure if it is the best approach.

larkox commented:
I added the "Create new subscription" step to the flow. I had to make some changes since this setting is not something that will be stored in the store. That brings me to some design doubts. The FlowStore cannot be implemented completely in this plugin implementation of the store, since we must be able to create the subscription. We could create a new store object inside the mscalendar, but it feel somehow wrong. The solution I chose (a implementation of FlowStore on the Welcomer bot) seems near the worst solution (but was fast to implement and deploy for you to see).

Any opinion on where should the FlowStore be implemented to allow these "not store related" updates? Maybe FlowStore is not the best name and should be also updated?

I also am thinking about removing refactoring the PostIDs to use something similar to a map, or even store it in the kvstore separated from the user.

larkox commented on server/utils/bot/flower.go:
```diff
@@ -0,0 +1,105 @@
+
+import "github.com/mattermost/mattermost-plugin-mscalendar/server/utils/flow"
+
+type Flower interface {
```
Since a "flower" controls the start, cancel, and follow the next step of all the flows, I think it is better to call it controller. And since the interface is outside the flow package, I decided to call it FlowController.

Nevertheless, I can move the interface to the flow package, and leave in the bot package only the implementation. That way the interface would be just "controller", and the file would keep being flow_controller.go

levb commented:
@larkox Should this now be marked as "Ready for Review"?

larkox commented:
@aaronrothschild Added 'empty steps' (just show a message) in particular for daily summary setting. Here the screenshot (will push in a few minutes).
![Screenshot from 2020-03-31 11-55-03](https://user-images.githubusercontent.com/1933730/78013346-a4528d80-7346-11ea-9bd8-58b24ff68f32.png)

DHaussermann commented:
Hi @larkox,
I've done a round of testing on this and have some follow-up questions... 
1. I cannot select an answer to the question `Do you want to receive notifications when you receive a new event?` both options throw a JS error `https://dkh-local.ngrok.io/api/v4/posts/5uni1g5th3deikxziqjxrz7q5c/actions/g1d9ubmnb38sxmi199sk11mskw 400 (Bad Request)` any thoughts? Also, this is the option I need to create a subscription correct?
2. Maybe outside the scope of this actual PR but, How do I get the  posts that show me the acceptance status? It't not in the daily summary. Possibly I need a working subscription (see 1. above)?

Minor UX issues:

3. When connecting, I don't see the new `Connect to your Mattermost Account?` button I still see the previous link instead.
 
4. Buttons for the affirmative choices don't have a blue background as shown in In-Vision

larkox commented:
> Hi @larkox,
> I've done a round of testing on this and have some follow-up questions...
> 
>     1. I cannot select an answer to the question `Do you want to receive notifications when you receive a new event?` both options throw a JS error `https://dkh-local.ngrok.io/api/v4/posts/5uni1g5th3deikxziqjxrz7q5c/actions/g1d9ubmnb38sxmi199sk11mskw 400 (Bad Request)` any thoughts? Also, this is the option I need to create a subscription correct?
Not sure why this is happening. Nevertheless, I just added new error handling so the error should be more clear. Would you mind checking again with the latest commit and tell me what does the error says?
> 
>     2. Maybe outside the scope of this actual PR but, How do I get the  posts that show me the acceptance status? It't not in the daily summary. Possibly I need a working subscription (see 1. above)?
That should not be in the scope of this PR. Daily Summary and calendar view should show that information, but the formatting is (if I am not mistaken) yet to be settle in stone. @mickmister could you confirm this?
> 
> 
> Minor UX issues:
> 
>     1. When connecting, I don't see the new `Connect to your Mattermost Account?` button I still see the previous link instead.
We still need to use the link due to the limitations we have in order to work on mobile (use normal messages or slack attachments). If we were to use an action button, it would not open the new window needed to do the whole oauth flow.
> 
>     2. Buttons for the affirmative choices don't have a blue background as shown in In-Vision
Due to the limitations we have in order to work on mobile (use normal messages or slack attachments) this is not possible as of today.

mickmister commented:
> Daily Summary and calendar view should show that information, but the formatting is (if I am not mistaken) yet to be settle in stone. @mickmister could you confirm this?

@larkox Yes, the markdown formatting for each kind of post is certainly subject to change. I think we should make sure all of them are in the `views` package, then converge on the format for each one.

We can post samples in the mscalendar plugin channel, and rally around improvements by posting markdown edits in the channel.

mickmister commented on server/utils/flow/handler.go:
```diff
@@ -0,0 +1,78 @@
+
+	property := step.GetPropertyName()
+
+	valueString := r.URL.Query().Get(property)
```
0/5 remove blank lines in this function like the one after the `property :=` line. I suggest joining it with the next block instead.

larkox commented on server/command/connect.go:
```diff
@@ -7,18 +7,30 @@ import (
 )
 
+const (
+	ConnectBotAlreadyConnectedTemplate = "The bot account is already connected to %s account `%s`. To connect to a different account, first run `/%s disconnect_bot`."
```
When the PR is merged, we can deal with the conflicts.

larkox commented on server/mscalendar/welcome_flow.go:
```diff
@@ -0,0 +1,85 @@
+		Title:                "Subscribe to events",
+		Message:              "Do you want to receive notifications when you receive a new event?",
+		PropertyName:         store.SubscribePropertyName,
+		TrueButtonMessage:    "Yes - I will like to receive notifications",
```
Right now there is no configuration for this. You always receive the notifications for upcoming events. On top of that, receiving notifications for upcoming events is still not on master, so any change on that side is prone to create new merge problems.

@aaronrothschild Make sense to give this configurability to the user? (whether they want to receive notifications for upcoming events or not)

mickmister replied:
I definitely think the feature warrants a config value, but that can be a separate ticket.

larkox commented on server/mscalendar/welcomer.go:
```diff
@@ -0,0 +1,184 @@
+const (
+	WelcomeMessage = `### Welcome to the Microsoft Calendar plugin!
+Here is some info to prove we got you logged in
+- Name: %s
```
This is legacy and can be removed, since it is not being used.

larkox commented on server/mscalendar/welcomer.go:
```diff
@@ -0,0 +1,184 @@
+- Name: %s
+`
+	ConnectSuccessTemplate = `Welcome to the Microsoft Calendar.
+	[Click here to link your account.](%s/oauth2/connect)`
```
Bad naming. Will fix.

larkox commented on server/mscalendar/welcomer.go:
```diff
@@ -0,0 +1,184 @@
+	return nil
+}
+
+func (bot *mscBot) AfterSuccessfullyConnect(userID, userLogin string) error {
+	postID, err := bot.Store.DeleteUserWelcomePost(userID)
+	if err != nil {
+		bot.Errorf("error deleting user welcom post id, err=" + err.Error())
+	}
+	if postID != "" {
+		post := &model.Post{
+			Id: postID,
+		}
+		model.ParseSlackAttachment(post, []*model.SlackAttachment{bot.getConnectedAttachment(userLogin)})
+		bot.DMUpdatePost(post)
+	}
```
Welcome post should always be present. Therefore, there will never be an error, and the post ID will never be empty. Nevertheless, unexpected things may happen for many reasons, for example, an admin deleting the bot post for who knows which reason. In order not to leave it in a bad state, these errors are silently logged, and continue the execution by starting the initial setting flow.

larkox commented on server/mscalendar/welcomer.go:
```diff
@@ -0,0 +1,184 @@
+func (bot *mscBot) getConnectAttachment() *model.SlackAttachment {
+	sa := model.SlackAttachment{
+		Title: "Connect",
+		Text:  fmt.Sprintf(ConnectSuccessTemplate, bot.pluginURL),
```
What about leaving it as WelcomeMessage? As the original welcome message will be deleted, we can use this.

larkox commented on server/mscalendar/welcomer.go:
```diff
@@ -0,0 +1,184 @@
+func (bot *mscBot) getConnectedAttachment(userLogin string) *model.SlackAttachment {
+	return &model.SlackAttachment{
+		Title: "Connect",
+		Text:  ":tada: Congratulations! Your microsoft account (*" + userLogin + "*) has been connected to Mattermost.",
```
I understand your pain, but I think that adding templates here is an overkill, specially with templates with just one element.
Also, the idea of leaving this text here is that it is just used here. The Welcome message is used on various places, and therefore the reason to leave it as a constant.

larkox commented on server/mscalendar/welcomer.go:
```diff
@@ -0,0 +1,184 @@
+	return nil
+}
+
+func (bot *mscBot) cleanPostIDs(mattermostUserID string) error {
+	postID, err := bot.Store.DeleteUserWelcomePost(mattermostUserID)
+	if err != nil {
+		return err
+	}
+
+	if postID != "" {
+		err = bot.DeletePost(postID)
```
The name is inherited from the first takes to this task. Will change.

larkox commented on server/mscalendar/welcomer.go:
```diff
@@ -0,0 +1,184 @@
+
+	if value {
+		mscalendar := New(bot.Env, userID)
+		_, err := mscalendar.CreateMyEventSubscription()
```
Yes, we will. If we do not follow the flow, call the slash command to subscribe, and then finish the flow, we will enter through this if. Shall we check if there is already an active subscription? Or what is your thought on this?

mickmister replied:
I don't see any reason why we would want the user to have two event subscriptions, so I believe we should check, and bail if it already exists.

larkox commented on server/utils/flow/handler.go:
```diff
@@ -0,0 +1,78 @@
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(response.ToJson())
+
+	fh.store.RemovePostID(mattermostUserID, property)
```
If we were to change this, for parallelism we should change also the GetPostID and SetPostID. Is this something you would like?

mickmister replied:
Nope, keeping it as-is seems fine :+1:

larkox commented on server/utils/bot/bot.go:
```diff
@@ -27,15 +30,28 @@ type bot struct {
+
+	flow      flow.Flow
+	flowStore flow.FlowStore
+	// TODO: Remove this and store this information in KVStore for HA
```
Already solved it.

mickmister commented on server/mscalendar/oauth2.go:
```diff
@@ -124,13 +120,14 @@ func (app *oauth2App) CompleteOAuth2(authedUserID, code, state string) error {
 		app.Poster.DM(authedUserID, BotWelcomeMessage, me.Mail)
 		return nil
+	} else {
+		app.Welcomer.AfterSuccessfullyConnect(mattermostUserID, me.Mail)
```
Still curious about this one

larkox replied:
I cannot remember nor find any reason, so I will put it after that.

mickmister commented on server/utils/settingspanel/handler.go:
```diff
@@ -29,33 +30,37 @@ func Init(h *httputils.Handler, panel Panel) {
-	sh.panel.Set(mattermostUserID, idString, value)
+	err := sh.panel.Set(mattermostUserID, idString, value)
+	if err != nil {
+		utils.SlackAttachmentError(w, "Error: cannot set the property, "+err.Error())
```
Do these errors get shown in the UI? It seems a bit low-level

larkox replied:
Yes, these would get shown in the UI. My idea is the error to be meaningful to debug, but we could move the internals of the error to a log in the server. Nevertheless, that would imply passing a logger to the handler, which will make things a bit dirtier.

DHaussermann commented:
@larkox I have done a test run through this now. Just a couple points.
1. It seems like the login workflow may not properly store the response to confirm login status change? Maybe this is just because the related PR is not in this branch?
![LoginWorkflow](https://user-images.githubusercontent.com/35588783/80008646-0447da00-8496-11ea-8dcc-d9d5c37147e6.PNG)

2. Using `mscalendar connect` creates the new post that is part of the login workflow as well as the previous one. So the user ends up with 2. Can you get rid of the old one?
![WorkflowConnect](https://user-images.githubusercontent.com/35588783/80009373-f6df1f80-8496-11ea-88bc-9ace1fb413c2.PNG)

3. UX point - Tell me HOW to change the settings here (using the slash command)
![LoginWorkflowSettings](https://user-images.githubusercontent.com/35588783/80008539-e0849400-8495-11ea-9112-d14bd82d9e39.PNG)

larkox commented:
1. That must be a bug. I will look into it ASAP.
2. It is a complicated decision. Slash commands must return something. And we want to have this connect message on the bot. At some point I thought about a "The MSCalendar bot just sent you the information to connect", but that adds another step. Also, the normal flow would be to call the command in a different channel than the bots direct message, so in general, you will not see the two messages together. @aaronrothschild Any thoughts on this?
3. My bad. This was changed at some point (since we didn't have a settings panel at the moment), and have to be changed again as you said.

aaronrothschild commented:
> 2\. It is a complicated decision. Slash commands must return something. And we want to have this connect message on the bot. At some point I thought about a "The MSCalendar bot just sent you the information to connect", but that adds another step. Also, the normal flow would be to call the command in a different channel than the bots direct message, so in general, you will not see the two messages together. @aaronrothschild Any thoughts on this?

@larkox sorry, I'm not following.  Where does the double message show up?

larkox commented:
@aaronrothschild 
After `/mscalendar connect` you will receive an answer to the slash command (and ephemeral message) on whichever channel you are. You will also receive a Direct Message from the bot. Both messages have almost the same information, with the only difference being that the bot one will change when you successfully connect.
If you were already on the Direct Message of the bot (something that normal users will not do in general), you will see both messages together, as the image Dylan shared.
We could remove the message from the bot and just leave the command response (the ephemeral message) provide the connection link, and when the connection is finished, then send the welcome message to the user.
Or we could give as a command response a message like "The MSCalendar bot just sent you the information on how to connect your account."

larkox commented:
@DHaussermann 
Just heads up that the latest commit should fix 1 and 3.

larkox commented:
@aaronrothschild I just discovered that slash commands can redirect the user to a different URL. What do you think about  `/mscalendar connect` redirects to the bot direct message channel? Also, my assumption about the slash commands not being able to return nothing may be wrong. The final experience would be:

I type `/mscalendar connect` and hit enter. The window changes automatically to the MSCalendar bot direct message. There I find the instructions to sign in.

larkox commented:
@DHaussermann Not sure what is happening with the CI, but this latest commit should fix also 2, so feel free to give it another round whenever you have the time.

DHaussermann commented:
@larkox Thanks for the fixes! 1., 2. and 3. are resolved.  

This may be ready to merge
- Tested answering questions in every applicable combination of yes/no. 
- Data saved is accurate 
- Confirmed 3 issues above are resolved

Just a couple points:
1. Intermittently I see the issue where answering the last question about subscription throws an error. This may be cause by data created by builds from other branches. Uninstalling the plugin and re-installing solves the issue even though it's the same build that was already running.
![ErrorOnSubscribe](https://user-images.githubusercontent.com/35588783/80262744-723dfe00-865c-11ea-8d8e-2dbc2d3e5e04.PNG)
Please advise if this needs to be investigated in this PR or we're okay creating a separate issue if we see it in master. 
 
2. I think we should implement this suggestion to redirect the user to where the post will be https://github.com/mattermost/mattermost-plugin-mscalendar/pull/44#issuecomment-618319673

larkox commented:
@DHaussermann 
1. It is weird. I think that this PR is becoming bit enough as it is, so I would say we merge this, and investigate the issue as a separate ticket.
2. This should be already implemented. Can you not see this behaviour in the latest build?

mickmister commented:
I've tested this as well with no issues found. Though I agree with Dylan's comment in our conversation on MM to make it so the user can change their mind within thr settings panel about whether or not they should receive event invitations.

> 1. It is weird. I think that this PR is becoming bit enough as it is, so I would say we merge this, and investigate the issue as a separate ticket.

Dylan and I have talked about stale data from different PRs affecting each other. I think it's important to document the behaviors so as to be on the lookout for them when testing master. 

Things like connecting to two different Microsoft accounts with one MM account (disconnecting in between them) may cause these sorts of things too. Things that we wouldn't directly design for. I'm not saying that caused this issue, just unexpected user actions like that may get the account in a weird state.

> 2. This should be already implemented. Can you not see this behavior in the latest build?

I can confirm this is implemented and working correctly. In fact, when running on localhost, it redirects me to my ngrok instance's bot DM channel. This initially caught me off guard, but it ensures I am logged into the necessary domain before connecting.

LGTM :+1:

@aaronrothschild We will need your blessing here

aaronrothschild commented:
Sweet!

mickmister commented:
/update-branch
